File: one_to_five_h/215-kth-largest-element-in-an-array.py
"""
    Leetcode 215
    https://leetcode-cn.com/problems/kth-largest-element-in-an-array/

"""


def findKthLargest(nums, k):
    if not nums:
        return None
    if len(nums) < 2:
        return nums[0]
    n = len(nums)
    build_heap(nums, k)
    for i in range(k, n):
        if nums[i] > nums[0]:
            nums[i], nums[0] = nums[0], nums[i]
            heapify(0, nums, k)
    return nums[0]


def build_heap(nums, n):
    for i in range((n - 1) // 2, -1, -1):
        heapify(i, nums, n)


def heapify(i, nums, n):
    left = 2 * i + 1
    right = 2 * i + 2
    maxi = i
    if left < n and nums[left] < nums[maxi]:
        maxi = left
    if right < n and nums[right] < nums[maxi]:
        maxi = right
    if maxi != i:
        nums[i], nums[maxi] = nums[maxi], nums[i]
        heapify(maxi, nums, n)

File: heap/215-kth-largest-element-in-an-array.go
package main

func findKthLargest(nums []int, k int) int {
	buildHeap(nums, k)
	for i := k; i < len(nums); i++ {
		if nums[i] > nums[0] {
			nums[i], nums[0] = nums[0], nums[i]
			heapify(nums, 0, k)
		}
	}
	return nums[0]
}

func buildHeap(nums []int, n int) {
	for i := n/2 - 1; i >= 0; i-- {
		heapify(nums, i, n)
	}
}
func heapify(nums []int, i, n int) {
	l, r := 2*i+1, 2*i+2
	minIdx := i
	if l < n && nums[l] < nums[minIdx] {
		minIdx = l
	}
	if r < n && nums[r] < nums[minIdx] {
		minIdx = r
	}
	if minIdx != i {
		nums[i], nums[minIdx] = nums[minIdx], nums[i]
		heapify(nums, minIdx, n)
	}
}
